libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+	case STATE_ADMINDOWN:
+		bs->bs_state = STATE_DOWN;
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
+	case STATE_DOWN:
+		/*
+		 * We are waiting peer's INIT.
+		 *
+		 * If we are passive, then we must manually respond to this
+		 * packet (TX timer is disabled).
+		 */
+		if (bs->bs_passive)
+			bfddp_send_control_packet(bs, arg);
+		break;
+	case STATE_INIT:
+		/* FALLTHROUGH. */
+	case STATE_UP:
+		bs->bs_state = STATE_UP;
+		bs->bs_diag = 0;
+
+		/* Start polling. */
+		bs->bs_poll = true;
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
+	}
+}
+
+static void
+bfddp_session_sm_up(struct bfd_session *bs, __attribute__((unused)) void *arg,
+		    enum bfd_state_value nstate)
+{
+	switch (nstate) {
+	case STATE_ADMINDOWN:
+		/* FALLTHROUGH. */
+	case STATE_DOWN:
+		bs->bs_state = STATE_DOWN;
+
+		bfddp_session_set_slowstart(bs);
+		/* Disable echo timers. */
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
+	case STATE_INIT:
+		/* FALLTHROUGH. */
+	case STATE_UP:
+		/* NOTHING. */
+		break;
+	}
+}
+
+void
+bfddp_session_state_machine(struct bfd_session *bs, void *arg,
+			    enum bfd_state_value nstate)
+{
+	enum bfd_state_value ostate = bs->bs_state;
+
+	switch (ostate) {
+	case STATE_ADMINDOWN:
+		bfddp_session_sm_admindown(bs, arg, nstate);
+		break;
+	case STATE_DOWN:
+		bfddp_session_sm_down(bs, arg, nstate);
+		break;
+	case STATE_INIT:
+		bfddp_session_sm_init(bs, arg, nstate);
+		break;
+	case STATE_UP:
+		bfddp_session_sm_up(bs, arg, nstate);
+		break;
+	}
+
+	/* Call application callback to notify about state change. */
+	if (ostate != bs->bs_state) {
+		/*
+		 * If state changed from UP to DOWN and we are in passive mode,
+		 * then * we need to disable the RX/TX timers to avoid sending
+		 * new control packets.
+		 */
+		if (bs->bs_passive
+		    && (ostate == STATE_UP && bs->bs_state == STATE_DOWN)) {
+			bfddp_callbacks.bc_rx_control_stop(bs, arg);
+			bfddp_callbacks.bc_tx_control_stop(bs, arg);
+		}
+
+		bfddp_callbacks.bc_state_change(bs, arg, ostate, bs->bs_state);
+	}
+}
+
+enum bfddp_packet_validation
+bfddp_session_validate_packet(const struct bfddp_control_packet *bcp,
+			      size_t bcplen)
+{
+	enum bfd_state_value state;
+	uint8_t version;
+
+	/* Assert we have the received the whole packet. */
+	if (bcplen < (int)sizeof(*bcp))
+		return BPV_PACKET_TOO_SMALL;
+
+	/* Check packet header length. */
+	if (bcp->length < sizeof(*bcp) || bcp->length > bcplen)
+		return BPV_INVALID_LENGTH;
+
+	/* Check version. */
+	version = (bcp->version_diag >> 5) & 0x07;
+	if (version != BFD_PROTOCOL_VERSION)
+		return BPV_INVALID_VERSION;
+
+	/* Invalid detection multiplier. */
+	if (bcp->detection_multiplier == 0)
+		return BPV_ZERO_MULTIPLIER;
+
+	/* Discard sessions using ID zero. */
+	if (bcp->local_id == 0)
+		return BPV_ZERO_LOCAL_ID;
+
+	/* Invalid remote ID with established session. */
+	state = (bcp->state_bits >> 6);
```

Comment:
For all state bits, might be worth defining `static inline` (or better `__attribute__((always_inline))`) or a MACRO to read the different flags. Might avoid some stupid bugs of shifting with numbers all over the codebase

Reply:
Sure, I'll add static inline functions in the header to do this.